Documentation/Plugin-developer-manual.md

```diff
@@ -0,0 +1,230 @@
+
+Modules dealing with client-specific data require a *SESSION* object for every client. As with the instance data, the definition of the module session structure is up to the writer and MaxScale treats it as an opaque type. Usually the session contains status indicators and any resources required by the client. MaxScale core has its own `MXS_SESSION` object, which tracks a variety of client related information. The `MXS_SESSION` is given as a parameter to module-specific session creation functions and is required for several typical operations such as connecting to backends.
+
+Descriptor control blocks (`DCB`), are generalized I/O descriptor types. DCBs store the file descriptor, state, remote address, username, session, and other data. DCBs are created whenever a new socket is created. Typically this happens when a new client connects or MaxScale connects the client session to backend servers. The module writer should use DCB handling functions provided by the CPI to manage connections instead of calling general networking libraries.
+
+Network data such as client queries and backend replies are held in a rather complicated buffer container type called `GWBUF`. Multiple GWBUFs can form a linked list with type information and properties in each GWBUF-node. Each node includes a pointer to a reference counted shared buffer (`SHARED_BUF`), which finally points to a slice of the actual data. What all this means is that multiple GWBUF-chains can share some parts of data while having some parts private. The construction is meant to minimize the need for data copying. In simple situations there is no need to form a linked list and the buffer acts similar to a handle. Plugin writers should use the CPI to manipulate GWBUFs.
+
+In the next sections, the APIs of module types in MariaDB MaxScale are presented.
+
+#### General module management
+
+```java
+int process_init()
+void process_finish()
+int thread_init()
+void thread_finish()
+```
+
+These four functions are present in all `MXS_MODULE` structs and are not part of the API of any individual module type. `process_init` and `process_finish` are called by the module loader right after loading a module and just before unloading a module. Usually, these can be set to null in `MXS_MODULE` unless the module needs some general initializations before creating any instances. `thread_init` and `thread_finish` are thread specific equivalents.
```

Comment:
Modules will actually not be unloaded. Perhaps better to say that `process_finish()` is called right before MaxScale terminates.